src/highlighter.js

```diff
@@ -237,24 +273,51 @@ export default function (options) {
                 if (currentNode.isSameNode(rangeInfos.startNode)) {
                     if (isText(rangeInfos.startNodeContainer) && rangeInfos.startOffset !== 0) {
                         // we defer the wrapping to the next iteration of the loop
+                        //end of node should be highlighted
                         rangeInfos.startNode = currentNode.splitText(rangeInfos.startOffset);
                         rangeInfos.startOffset = 0;
+                        splitDatas.push({ node: rangeInfos.startNode, beforeWasSplit: true, afterWasSplit: false });
                     } else {
                         isWrapping = true;
                     }
                 }
 
                 if (currentNode.isSameNode(rangeInfos.endNode) && isText(rangeInfos.endNodeContainer)) {
                     if (rangeInfos.endOffset !== 0) {
-                        currentNode.splitText(rangeInfos.endOffset);
+                        if (rangeInfos.endOffset < currentNode.textContent.length) {
+                            //start of node should be highlighted
+                            currentNode.splitText(rangeInfos.endOffset);
+                            splitDatas.push({ node: currentNode, beforeWasSplit: false, afterWasSplit: true });
+                        } else {
+                            //whole node should be highlighted
+                            splitDatas.push({ node: currentNode, beforeWasSplit: false, afterWasSplit: false });
+                        }
                     } else {
                         isWrapping = false;
                     }
                 }
 
                 // wrap the current node...
                 if (isText(currentNode)) {
-                    wrapTextNode(currentNode, currentGroupId);
+                    if (!keepEmptyNodes) {
+                        wrapTextNode(currentNode, currentGroupId);
+                    } else {
+                        //do the work of `unwrapEmptyHighlights` ahead of time, before `mergeAdjacentNodes` runs,
+                        //because in this branch we do not want to add nodes to dom unless necessary.
+                        //also be more strict and don't allow to select nodes with spaces only, because the may appear in unexpected places in markup
+                        if (currentNode.textContent.length && (
+                                currentNode.textContent.trim().length ||
+                                (currentNode.previousSibling && isWrappingNode(currentNode.previousSibling) && currentNode.previousSibling.className === className) ||
+                                (currentNode.nextSibling && isWrappingNode(currentNode.nextSibling) && currentNode.nextSibling.className === className))) {
```

Comment:
The condition is pretty hard to read. Is it possible to extract some of the inner parts to named variables?